go/k8s/controllers/user_controller.go

```diff
@@ -0,0 +1,235 @@
+	DBProxyURL         string `json:"dbProxyUrl"`
+	DBProxyAccessToken string `json:"dbProxyAccessToken"`
+	HumanName          string `json:"humanName"`
+	Email              string `json:"email"`
+}
+
+func platformOptions(user *sql.User, developerKey string, dbProxyURL string, dbProxyToken string) *PlatformOptions {
+	return &PlatformOptions{
+		UserID:             user.ID,
+		CloudID:            user.CloudID,
+		AuthToken:          user.AuthToken,
+		DeveloperKey:       developerKey,
+		Locale:             user.Locale,
+		Timezone:           user.Timezone,
+		DBProxyURL:         dbProxyURL,
+		DBProxyAccessToken: dbProxyToken,
+		HumanName:          user.HumanName,
+		Email:              user.Email,
+	}
+}
+
+// JSONResposne from http service
+type JSONResponse struct {
+	Result string      `json:"result"`
+	Data   interface{} `json:"data"`
+}
+
+//+kubebuilder:rbac:groups=backend.almond.stanford.edu,resources=users,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=endpoints,verbs=get;list;watch
+//+kubebuilder:rbac:groups=backend.almond.stanford.edu,resources=users/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=backend.almond.stanford.edu,resources=users/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the User object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *UserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	backendURL, userID, err := r.getBackendURL(req.NamespacedName.Name, ctx, req)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	userState, err := r.engineStatus(userID, backendURL)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	user := &backendv1.User{}
+	if err := r.Client.Get(ctx, req.NamespacedName, user); err != nil {
+		if apierrors.IsNotFound(err) {
+			// User is already deleted, kill engine if it's still running.
+			if userState == Running {
+				return ctrl.Result{}, r.killEngine(userID, backendURL)
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if !user.ObjectMeta.DeletionTimestamp.IsZero() {
+		// object is marked for deletion
+		if userState == Running {
+			return ctrl.Result{}, r.killEngine(userID, backendURL)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	user.Status.Backend = backendURL
+	defer r.Status().Update(ctx, user)
+	if userState == Running {
+		user.Status.State = string(Running)
+		return ctrl.Result{RequeueAfter: 20 * time.Second}, nil
+	}
+
+	if err = r.runEngine(user.Spec.ID, backendURL); err != nil {
```

Comment:
Do we also need to create the worker pods here? Or at least select which worker to run into?

Are worker pods controlled by a deployment, statefulset, or by this controller?

Reply:
For shared mode,  it's load balance among the statefulset backends.  The statefulset is setup by manually. The the future, we can add autoscaling to the statefulset either in the contorller or through a monitoring job.